### PR TITLE
feat: Feature/evaluated value window

### LINKED ISSUE #4458: [Bug]Evaluated value popup overlaps the DB query editor when the browser is minimized
## Description
Release Environment - PostgresSQL - It is observed that due to overlapping the cursor in the DB query editor is hidden due to which the user feels he/she cannot complete the action 

Expected Result:
Cursor should be visible clearly in the minimized mode and the evaluated value popup should be placed appropriately 

Actual Result:
Due to overlapping the cursor in the DB query editor is hidden due to which the user feels he/she cannot complete the action 

### Steps to reproduce the behaviour:
1. Launch Appmsith app in the Release Environment 
2. Expand DB
3. Create new Datasource with postgresSQL
4. Set up authentication 
5. Minimise the screen to 70-80% mode 
6. Check for the Evaluated value popup overlap and the DB query Cursor 

**VideoLink:**
[LOOM DEMO](https://www.loom.com/share/805ca00592614bdb8f4b02535927e760)

### Important Details
 - Version: [Chrome - 90.0.4430.212 ]**
 - OS: [MacOS Big Sur 11.3]
 - Browser [Chrome]

### PR DESCRIPTION
**Pull request by Irongade**
Evaluated value window improvements:

1. Evaluated value window will not show up unless there is a dynamic binding within the input field or when there's a validation error on the input field (this edge case happens when user types an incorrect string in the field for example typing a wrong date format inside the Default Date field in the DatePickerWidget).
2. Evaluated value window will be draggable anywhere on the DOM. The new position would also be remembered the next time the user returns to the input.
3. The UI of the evaluated value window has been changed: Since the popup will become draggable, it will have the name of the input it is attached to. The styling of the evaluated value window fields has been changed as well.

Design Documentation: https://www.notion.so/appsmith/Improve-the-evaluated-value-window-experience-fbbdead747a5468b8015c2fbbb5843bc

> Add a TL;DR when description is extra long (helps content team)

Fixes #4458 

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

> Tests

- Manual
- Cypress

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Autocomplete/autocomplete_spec.js [hunk 1]
```diff
@@ -78,15 +78,4 @@ describe("Dynamic input autocomplete", () => {
...
-  it("opens current value popup", () => {
-    // Test on api pane
-    cy.NavigateToAPI_Panel();
-    cy.get(apiwidget.createapi).click({ force: true });
-    cy.wait("@createNewApi");
-    cy.get(apiwidget.headerValue).within(() => {
-      cy.get("textarea").click({ force: true });
-    });
-    cy.assertEvaluatedValuePopup("string");
-  });
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Autocomplete/autocomplete_spec.js:91]: Why is this case removed?

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Button/Button_spec.js [hunk 2]
```diff
@@ -86,12 +86,6 @@ describe("Button Widget Functionality", function() {
 
     //Changing the text on the Button
     cy.testJsontext("label", this.data.ButtonLabel);
-    cy.get(commonlocators.evaluatedTypeTitle)
-      .first()
-      .find("span")
-      .click();
-    cy.EvaluateDataType("string");
-    cy.EvaluateCurrentValue(this.data.ButtonLabel);
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Button/Button_spec.js:94]: Why is this removed?

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts [hunk 3]
```diff
@@ -38,7 +38,7 @@ describe("Category Slider spec", () => {
 
     propPane.UpdatePropertyFieldValue("Default Value", "md");
 
-    agHelper.VerifyEvaluatedValue("md");
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts:41]: Why is this validation removed?

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Autocomplete/autocomplete_spec.js (hunk 1 above)
Comment (Irongade) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Autocomplete/autocomplete_spec.js:91]: One of the changes made to the evaluated value pop up is for the pop up not to be visible if the user has not made use of dynamic bindings within the field. Hence, since most of the cypress tests make use of plain strings for testing or in this case by simply clicking the input field, it would no longer be possible to assert that the pop up has any values or show up (become visible), simply because it has been intentionally disabled. The same goes for most other cypress tests that deal with using  the evaluated value popup to assert what the user has typed into the input field or to assert the expected values and expected examples.

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Button/Button_spec.js (hunk 2 above)
Comment (Irongade) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Button/Button_spec.js:94]: Here we are asserting the label field of the button widget is a string value of "Test label", hence the evaluated pop up will not be visible, hence no need to assert evaluated datatype and current value

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts (hunk 3 above)
Comment (Aishwarya-U-R) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts:41]: Can we add the above explanation to the method's definition - so if someone wants to call the method in future will know they need to validate it for dynamic binding like `{{Api1.data}}` only.

File: app/client/cypress/integration/Regression_TestSuite/ServerSideTests/QueryPane/EvaluatedValuePopUp_spec.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import { ObjectsRegistry } from "../../../../support/Objects/Registry";
+import formControls from "../../../../locators/FormControl.json";
+
+const agHelper = ObjectsRegistry.AggregateHelper,
+  dataSources = ObjectsRegistry.DataSources;
+
+describe("Ensures evaluated popup is viewable when dynamic bindings are present and draggable", function() {
+  beforeEach(() => {
+    dataSources.StartDataSourceRoutes();
+  });
```
Comment (Aishwarya-U-R) [inline app/client/cypress/integration/Regression_TestSuite/ServerSideTests/QueryPane/EvaluatedValuePopUp_spec.ts:10]: Is not needed as  already started in index.js beforeeach. Please remove

File: app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts (hunk 3 above)
Comment (Irongade) [reply app/client/cypress/integration/Regression_TestSuite/ClientSideTests/Widgets/Sliders/CategroySlider_spec.ts:41]: Yes, I will add a comment there

File: app/client/cypress/support/Pages/AggregateHelper.ts [hunk 5]
```diff
@@ -869,9 +944,7 @@ export class AggregateHelper {
       });
   }
 
-  public EvaluateExistingPropertyFieldValue(fieldName = "", currentValue = "") {
-    let toValidate = false;
-    if (currentValue) toValidate = true;
+  public EvaluateExistingPropertyFieldValue(fieldName = "") {
```
Comment (Aishwarya-U-R) [inline app/client/cypress/support/Pages/AggregateHelper.ts:947]: Why is currentValue & its validation removed? The purpose of it is to evaluate the existing property field value of the passed field as specified in the method name.

Comment (Irongade) [reply app/client/cypress/support/Pages/AggregateHelper.ts:947]: The reason that was removed is still for the reason I mentioned above, we were using the evaluated value popup to validate the current value, which is no longer valid. Hence why I removed it, Instead what i did is to return the value of the input and wrap it using `cy.wrap` which can then be chained with `.then` and used in the cypress test file. Kindly observe the changes.

Comment (Aishwarya-U-R) [reply app/client/cypress/support/Pages/AggregateHelper.ts:947]: I did go thru the changes. The whole point of this method was the Evaluate the existing value. If it is the text that is sent to be validated - it should be the caller that is to be changed & not this method. 
The purpose of this method is to evaluate the currentValue sent. The value sent from caller could also be a dynamic binding value.
Hence, revert the changes to this method to as is.

File: app/client/src/pages/Editor/Popper.tsx [hunk 6]
```diff
@@ -57,47 +65,89 @@ const DragHandleBlock = styled.div`
...
+      contentRef.current.style.top = e.top + "px";
+      contentRef.current.style.left = e.left + "px";
+
+      props.setPosition(e);
+
+      // add focus back to codemirror component.
+      if (
+        props?.editorRef &&
+        props?.editorRef?.current &&
+        props?.editorRef?.current?.children[1] &&
```
[the remark this answers is not in the document]
Comment (Irongade) [reply app/client/src/pages/Editor/Popper.tsx:99]: This is because we are traversing the child divs of the editor component reference to get the Codemirror component so we can call the focus method on it, and it so happens the Codemirror component is nested in the second child of the editor ref component (index 1).

[the remark this answers is not in the document]
Comment (sneha122) [reply app/client/src/pages/Editor/Popper.tsx:99]: Okay got it @Irongade Thanks!

**Review by sneha122 — DISMISSED** [review]

**Review by sneha122 — DISMISSED** [review]

**Review by sneha122 — APPROVED** [review]
